File: MotoRush/js/track.js
// MotoRush — procedural track generator.
// Produces a 1D heightfield in metres plus feature metadata used by the
// renderer (props, banners, shortcuts) and the AI (lookahead planning).

import { BIOMES } from './data.js';

export function mulberry32(seed) {
  let a = seed >>> 0;
  return function () {
    a |= 0; a = (a + 0x6D2B79F5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export const STEP = 0.4; // metres between height samples

const lerp = (a, b, t) => a + (b - a) * t;
const smooth = (t) => t * t * (3 - 2 * t);

// ——— Feature builders ———————————————————————————————————————————————
// Each builder appends samples to `pts` (absolute heights) and returns the
// feature descriptor so the AI/renderer know what is coming.

function pushRamp(pts, base, len, rise, shape = 'smooth') {
  const n = Math.max(2, Math.round(len / STEP));
  for (let i = 1; i <= n; i++) {
    const t = i / n;
    const k = shape === 'smooth' ? smooth(t) : shape === 'kicker' ? Math.pow(t, 1.3) : t;
    pts.push(base + rise * k);
  }
  return base + rise;
}

function pushFlat(pts, base, len) {
  const n = Math.max(1, Math.round(len / STEP));
  for (let i = 0; i < n; i++) pts.push(base);
  return base;
}

function pushWhoops(pts, base, count, amp, spacing) {
  const total = count * spacing;
  const n = Math.round(total / STEP);
  for (let i = 1; i <= n; i++) {
    const x = i * STEP;
    pts.push(base + amp * (1 - Math.cos((x / spacing) * Math.PI * 2)) * 0.5);
  }
  return base;
}

function pushBerm(pts, base, len, depth) {
  // A banked, scooped turn read side-on as a compression the rider can pump.
  const n = Math.round(len / STEP);
  for (let i = 1; i <= n; i++) {
    const t = i / n;
    pts.push(base - depth * Math.sin(t * Math.PI));
  }
  return base;
}

function pushRollers(pts, base, count, amp, spacing) {
  const n = Math.round((count * spacing) / STEP);
  for (let i = 1; i <= n; i++) {
    const x = i * STEP;
    pts.push(base + amp * Math.sin((x / spacing) * Math.PI * 2) * 0.5);
  }
  return base;
}

// ——— Generator ————————————————————————————————————————————————————
export function generateTrack(opts = {}) {
  const seed = opts.seed ?? (Math.random() * 1e9) | 0;
  const rnd = mulberry32(seed);
  const biomeKey = opts.biome && BIOMES[opts.biome] ? opts.biome : 'motocross';
  const biome = BIOMES[biomeKey];
  const difficulty = Math.min(1, Math.max(0, opts.difficulty ?? 0.5));
  const targetLength = opts.length ?? lerp(600, 1200, difficulty) * lerp(0.85, 1.15, rnd());

  const pts = [0];
  const features = [];
  const props = [];
  const shortcuts = [];
  let base = 0;
  let x = 0;

  // Start straight — long enough for a gate drop and first-gear pull.
  base = pushFlat(pts, base, 22);
  x = (pts.length - 1) * STEP;

  const jumpBias = biome.jumpBias * lerp(0.7, 1.35, difficulty);
  const whoopBias = biome.whoopBias * lerp(0.6, 1.3, difficulty);

  const weightedPick = () => {
    const table = [
      ['tabletop', 1.0 * jumpBias],
      ['double', 0.85 * jumpBias],
      ['triple', 0.45 * jumpBias * difficulty * 1.6],
      ['stepup', 0.6 * jumpBias],
      ['stepdown', 0.55 * jumpBias],
      ['whoops', 0.9 * whoopBias],
      ['rollers', 0.8],
      ['berm', 0.9],
      ['rhythm', 0.7 * jumpBias],
      ['straight', 0.5],
      ['hill', 0.7 * biome.rolling]
    ];
    let total = 0; for (const [, w] of table) total += w;
    let r = rnd() * total;
    for (const [k, w] of table) { r -= w; if (r <= 0) return k; }
    return 'straight';
  };

  let lastKind = '';
  while (x < targetLength) {
    let kind = weightedPick();
    if (kind === lastKind && rnd() < 0.7) kind = weightedPick();
    lastKind = kind;
    const startX = x;
    const sev = lerp(0.7, 1.35, difficulty) * lerp(0.85, 1.15, rnd());

    switch (kind) {
      case 'tabletop': {
        const h = lerp(1.1, 2.1, rnd()) * sev;
        const face = lerp(6.5, 9.5, rnd());
        base = pushRamp(pts, base, face, h, 'kicker');
        base = pushFlat(pts, base, lerp(5, 12, rnd()) * sev);
        base = pushRamp(pts, base, face * 2.2, -h, 'smooth');
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, height: h, risk: 0.35 });
        break;
      }
      case 'double': {
        const h = lerp(1.2, 2.0, rnd()) * sev;
        const gap = lerp(6, 10, rnd()) * sev;
        base = pushRamp(pts, base, lerp(6, 8, rnd()), h, 'kicker');
        const top = base;
        base = pushRamp(pts, base, 2.2, -h * 0.95, 'linear');
        base = pushFlat(pts, base, gap);
        base = pushRamp(pts, base, lerp(4, 6, rnd()), h * 0.95, 'smooth');
        base = pushRamp(pts, base, lerp(11, 15, rnd()), -h, 'smooth');
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, height: h, gap, top, risk: 0.65 });
        break;
      }
      case 'triple': {
        const h = lerp(1.5, 2.4, rnd()) * sev;
        base = pushRamp(pts, base, 7.5, h, 'kicker');
        base = pushRamp(pts, base, 2.4, -h, 'linear');
        base = pushFlat(pts, base, lerp(8, 11, rnd()));
        base = pushRamp(pts, base, 4, h * 0.7, 'smooth');
        base = pushRamp(pts, base, 6.5, -h * 0.7, 'smooth');
        base = pushFlat(pts, base, lerp(8, 11, rnd()));
        base = pushRamp(pts, base, 4.5, h * 0.9, 'smooth');
        base = pushRamp(pts, base, 14, -h * 0.9, 'smooth');
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, height: h, risk: 0.9 });
        // Rolling the triple is the safe line; sending it is the shortcut.
        shortcuts.push({ x: startX, len: (pts.length - 1) * STEP - startX, gain: 1.4, hint: 'Send the triple' });
        break;
      }
      case 'stepup': {
        const h = lerp(1.4, 2.6, rnd()) * sev;
        base = pushRamp(pts, base, lerp(6.5, 8.5, rnd()), h * 0.6, 'kicker');
        base = pushRamp(pts, base, 2, -h * 0.15, 'linear');
        base = pushFlat(pts, base, lerp(5, 9, rnd()));
        base = pushRamp(pts, base, 4, h * 0.55, 'smooth');
        base = pushFlat(pts, base, 6);
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, height: h, risk: 0.6 });
        break;
      }
      case 'stepdown': {
        const h = lerp(1.3, 2.4, rnd()) * sev;
        base = pushRamp(pts, base, 7, h * 0.5, 'kicker');
        base = pushRamp(pts, base, 1.6, -h * 0.1, 'linear');
        base = pushFlat(pts, base, lerp(6, 10, rnd()));
        base = pushRamp(pts, base, 11, -h, 'smooth');
        base = pushFlat(pts, base, 5);
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, height: h, risk: 0.6 });
        break;
      }
      case 'whoops': {
        const count = Math.round(lerp(5, 11, rnd()) * lerp(0.8, 1.3, difficulty));
        const amp = lerp(0.5, 1.0, rnd()) * sev;
        const spacing = lerp(3.4, 4.6, rnd());
        pushWhoops(pts, base, count, amp, spacing);
        features.push({ kind, x: startX, len: count * spacing, amp, spacing, risk: 0.7 });
        break;
      }
      case 'rollers': {
        const count = Math.round(lerp(3, 6, rnd()));
        pushRollers(pts, base, count, lerp(0.6, 1.4, rnd()) * sev, lerp(6, 9, rnd()));
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, risk: 0.25 });
        break;
      }
      case 'berm': {
        pushBerm(pts, base, lerp(10, 18, rnd()), lerp(0.8, 1.8, rnd()) * sev);
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, risk: 0.2 });
        break;
      }
      case 'rhythm': {
        const n = Math.round(lerp(3, 5, rnd()));
        const h = lerp(0.8, 1.5, rnd()) * sev;
        for (let i = 0; i < n; i++) {
          base = pushRamp(pts, base, lerp(4.5, 6, rnd()), h, 'kicker');
          base = pushRamp(pts, base, lerp(7, 10, rnd()), -h, 'smooth');
          base = pushFlat(pts, base, lerp(1.5, 4, rnd()));
        }
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, count: n, risk: 0.8 });
        shortcuts.push({ x: startX, len: (pts.length - 1) * STEP - startX, gain: 1.2, hint: 'Blitz the rhythm' });
        break;
      }
      case 'hill': {
        const h = lerp(3, 9, rnd()) * biome.rolling * 0.6;
        const up = rnd() < 0.5 ? 1 : -1;
        base = pushRamp(pts, base, lerp(14, 26, rnd()), h * up, 'smooth');
        features.push({ kind, x: startX, len: (pts.length - 1) * STEP - startX, risk: 0.15 });
        break;
      }
      default: {
        base = pushFlat(pts, base, lerp(10, 22, rnd()));
        features.push({ kind: 'straight', x: startX, len: (pts.length - 1) * STEP - startX, risk: 0.05 });
      }
    }

    // Gentle drift keeps long tracks from becoming a flat ribbon.
    base += (rnd() - 0.5) * 0.6 * biome.rolling;
    base = pushFlat(pts, base, lerp(2, 6, rnd()));
    x = (pts.length - 1) * STEP;
  }

  // Finish straight + landing runoff.
  base = pushFlat(pts, base, 40);

  // Light smoothing pass removes sample-level discontinuities that would
  // otherwise read as invisible curbs to the suspension solver.
  for (let pass = 0; pass < 2; pass++) {
    for (let i = 1; i < pts.length - 1; i++) {
      pts[i] = pts[i] * 0.6 + (pts[i - 1] + pts[i + 1]) * 0.2;
    }
  }

  const length = (pts.length - 1) * STEP;

  // Scenery props (purely decorative, deterministic from the seed).
  for (let i = 0; i < length; i += 6) {
    if (rnd() < 0.35) {
      props.push({
        x: i + rnd() * 4,
        type: biome.trees ? 'tree' : biome.stadium ? 'banner' : rnd() < 0.4 ? 'flag' : 'bale',
        scale: lerp(0.7, 1.4, rnd()),
        layer: rnd() < 0.5 ? 0 : 1
      });
    }
  }

  const startX = 6;
  const finishX = length - 26;

  return {
    seed, biomeKey, biome, difficulty, pts, step: STEP, length,
    features, props, shortcuts, startX, finishX,
    name: trackName(rnd, biome),
    heightAt(px) {
      const f = px / STEP;
      if (f <= 0) return pts[0];
      if (f >= pts.length - 1) return pts[pts.length - 1];
      const i = f | 0;
      return lerp(pts[i], pts[i + 1], f - i);
    },
    slopeAt(px) {
      const h1 = this.heightAt(px - STEP), h2 = this.heightAt(px + STEP);
      return (h2 - h1) / (2 * STEP);
    },
    featureAt(px) {
      for (const f of this.features) if (px >= f.x && px < f.x + f.len) return f;
      return null;
    }
  };
}

const PREFIX = ['Copper', 'Iron', 'Hollow', 'Wild', 'Broken', 'Silver', 'Red', 'Black', 'High',
  'Thunder', 'Dust', 'Sky', 'Storm', 'Cinder', 'Glass', 'Pine', 'Salt', 'Ghost'];
const SUFFIX = ['Ridge', 'Basin', 'Creek', 'Flats', 'Canyon', 'Hollow', 'Bowl', 'Pass', 'Sands',
  'Bluff', 'Gulch', 'Park', 'Valley', 'Point', 'Rise'];

function trackName(rnd, biome) {
  const a = PREFIX[(rnd() * PREFIX.length) | 0];
  const b = SUFFIX[(rnd() * SUFFIX.length) | 0];
  return `${a} ${b}`;
}

// Deterministic "hundreds of tracks" catalogue: a seeded index the career and
// quick-race screens page through. Any seed reproduces the exact same track.
export function trackCatalogue(count = 240) {
  const rnd = mulberry32(0xC0FFEE);
  const keys = Object.keys(BIOMES);
  const list = [];
  for (let i = 0; i < count; i++) {
    const biome = keys[(rnd() * keys.length) | 0];
    const difficulty = Math.min(0.98, 0.15 + (i / count) * 0.8 + (rnd() - 0.5) * 0.12);
    list.push({ id: i, seed: (rnd() * 1e9) | 0, biome, difficulty });
  }
  return list;
}

File: MotoRushiOS/MotoRush/Track.swift
import Foundation

// MARK: - Deterministic RNG

/// mulberry32 — small, fast, and identical across runs for a given seed, which
/// is what lets every track in the game be reproduced from a single Int.
struct RNG {
    private var state: UInt32
    init(seed: UInt32) { state = seed }

    mutating func next() -> Double {
        state = state &+ 0x6D2B79F5
        var t = state
        t = (t ^ (t >> 15)) &* (1 | t)
        t = t ^ (t &+ ((t ^ (t >> 7)) &* (61 | t)))
        return Double((t ^ (t >> 14))) / 4294967296.0
    }

    mutating func range(_ a: Double, _ b: Double) -> Double { a + (b - a) * next() }
    mutating func int(_ n: Int) -> Int { min(n - 1, Int(next() * Double(n))) }
}

func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double { a + (b - a) * t }
func clampd(_ v: Double, _ a: Double, _ b: Double) -> Double { v < a ? a : (v > b ? b : v) }

// MARK: - Biomes

struct Biome {
    let key: String
    let name: String
    let skyTop: String
    let skyBottom: String
    let ground: String
    let groundDeep: String
    let accent: String
    let dust: String
    let rolling: Double
    let jumpBias: Double
    let whoopBias: Double
    let grip: Double
    let rollDrag: Double
    let night: Bool
    let stadium: Bool
    let weather: String   // "", "rain", "snow", "embers"

    static let all: [Biome] = [
        Biome(key: "motocross", name: "Motocross", skyTop: "#7EC8FF", skyBottom: "#DFF1FF",
              ground: "#8A6244", groundDeep: "#5D4230", accent: "#C8A97A", dust: "#C9A97E",
              rolling: 1.0, jumpBias: 1.0, whoopBias: 0.8, grip: 1.0, rollDrag: 1.0,
              night: false, stadium: false, weather: ""),
        Biome(key: "supercross", name: "Supercross", skyTop: "#1B2340", skyBottom: "#38406B",
              ground: "#6B4C37", groundDeep: "#3D2B1F", accent: "#9C7A55", dust: "#B39A76",
              rolling: 0.4, jumpBias: 1.6, whoopBias: 1.6, grip: 1.05, rollDrag: 1.0,
              night: true, stadium: true, weather: ""),
        Biome(key: "sand", name: "Sand", skyTop: "#FFD79A", skyBottom: "#FFECCD",
              ground: "#E0C081", groundDeep: "#B3945C", accent: "#F0DCAE", dust: "#EFD9A8",
              rolling: 1.4, jumpBias: 0.7, whoopBias: 1.8, grip: 0.78, rollDrag: 2.1,
              night: false, stadium: false, weather: ""),
        Biome(key: "mud", name: "Mud", skyTop: "#7F8794", skyBottom: "#B9C0C9",
              ground: "#4F4034", groundDeep: "#332A22", accent: "#6A5646", dust: "#5A4A3B",
              rolling: 1.1, jumpBias: 0.9, whoopBias: 1.0, grip: 0.7, rollDrag: 1.7,
              night: false, stadium: false, weather: "rain"),
        Biome(key: "forest", name: "Forest", skyTop: "#8FD3A5", skyBottom: "#E6F7EA",
              ground: "#5C4A32", groundDeep: "#3A2F20", accent: "#6F8F4F", dust: "#8A7A58",
              rolling: 1.2, jumpBias: 0.9, whoopBias: 0.9, grip: 1.02, rollDrag: 1.0,
              night: false, stadium: false, weather: ""),
        Biome(key: "snow", name: "Snow", skyTop: "#C9DCEF", skyBottom: "#F6FBFF",
              ground: "#E8F0F7", groundDeep: "#A9BDD0", accent: "#FFFFFF", dust: "#FFFFFF",
              rolling: 1.0, jumpBias: 0.8, whoopBias: 0.9, grip: 0.62, rollDrag: 1.3,
              night: false, stadium: false, weather: "snow"),
        Biome(key: "desert", name: "Desert", skyTop: "#FFB27A", skyBottom: "#FFE3C2",
              ground: "#C98F5C", groundDeep: "#8E6039", accent: "#E6B98A", dust: "#E3C093",
              rolling: 1.5, jumpBias: 1.1, whoopBias: 0.7, grip: 0.9, rollDrag: 1.1,
              night: false, stadium: false, weather: ""),
        Biome(key: "night", name: "Night", skyTop: "#0B1026", skyBottom: "#1D2748",
              ground: "#3B3145", groundDeep: "#241E2C", accent: "#5A4C68", dust: "#6B5F78",
              rolling: 1.0, jumpBias: 1.2, whoopBias: 1.1, grip: 0.95, rollDrag: 1.0,
              night: true, stadium: false, weather: ""),
        Biome(key: "mountain", name: "Mountain", skyTop: "#6FA8DC", skyBottom: "#DBE9F7",
              ground: "#6D6A63", groundDeep: "#43413C", accent: "#8D8A80", dust: "#9C968A",
              rolling: 1.8, jumpBias: 1.2, whoopBias: 0.6, grip: 0.98, rollDrag: 1.0,
              night: false, stadium: false, weather: ""),
        Biome(key: "beach", name: "Beach", skyTop: "#5FC9FF", skyBottom: "#D8F3FF",
              ground: "#DCC394", groundDeep: "#A98F63", accent: "#F2E2BD", dust: "#EDDCB4",
              rolling: 0.9, jumpBias: 0.9, whoopBias: 1.2, grip: 0.85, rollDrag: 1.4,
              night: false, stadium: false, weather: ""),
        Biome(key: "volcano", name: "Volcano", skyTop: "#3A1020", skyBottom: "#7B2531",
              ground: "#3A2B2B", groundDeep: "#1D1414", accent: "#FF6A3D", dust: "#8C5A45",
              rolling: 1.3, jumpBias: 1.4, whoopBias: 1.0, grip: 0.92, rollDrag: 1.0,
              night: true, stadium: false, weather: "embers"),
        Biome(key: "arena", name: "Arena", skyTop: "#141A2E", skyBottom: "#2A3352",
              ground: "#71513A", groundDeep: "#3F2C20", accent: "#A78156", dust: "#B39A76",
              rolling: 0.3, jumpBias: 1.8, whoopBias: 1.5, grip: 1.06, rollDrag: 1.0,
              night: true, stadium: true, weather: "")
    ]

    static func named(_ key: String) -> Biome {
        return all.first(where: { $0.key == key }) ?? all[0]
    }
}

// MARK: - Track

struct TrackFeature {
    let kind: String
    let x: Double
    let len: Double
    let risk: Double
}

struct Prop {
    let x: Double
    let kind: String      // tree, bale, flag, banner
    let scale: Double
    let back: Bool
}

/// A track is a 1D heightfield in metres plus the feature metadata the AI and
/// the renderer need. Everything is derived from `seed`, so a track id is all
/// that has to be stored or sent over a wire.
final class Track {
    static let step: Double = 0.4

    // Geometry the generator is not allowed to exceed. The severity multiplier
    // used to scale features without any bound on the *slope* that resulted — a
    // "roller" could come out at 49 degrees, and terrain climbing under the
    // wheel that fast launches the bike far beyond anything a rider could do.
    static let maxFaceSlope: Double = 0.62    // ~32 degrees, whoops/rollers
    static let maxLaunchSlope: Double = 0.45  // ~24 degrees, steepest jump lip

    let seed: UInt32
    let biome: Biome
    let difficulty: Double
    private(set) var pts: [Double]
    let features: [TrackFeature]
    let props: [Prop]
    let length: Double
    let startX: Double
    let finishX: Double
    let name: String

    init(seed: UInt32, biomeKey: String, difficulty: Double, lengthOverride: Double? = nil) {
        self.seed = seed
        self.biome = Biome.named(biomeKey)
        self.difficulty = clampd(difficulty, 0, 1)
        var rnd = RNG(seed: seed)
        let d = self.difficulty
        let target = lengthOverride ?? (lerp(520, 1050, d) * rnd.range(0.85, 1.15))

        var pts: [Double] = [0]
        var features: [TrackFeature] = []
        var base: Double = 0

        func pushFlat(_ len: Double) {
            let n = max(1, Int((len / Track.step).rounded()))
            for _ in 0..<n { pts.append(base) }
        }
        func pushRamp(_ len: Double, _ rise: Double, _ shape: String) {
            // How steeply each shape exits at its top. `smooth` flattens out, so
            // it never launches; the other two do. Lengthen a takeoff rather
            // than shortening it, which is what a real track crew does: the jump
            // keeps its height, the lip just stops being a launcher.
            var len = len
            let factor: Double = shape == "kicker" ? 1.3 : (shape == "linear" ? 1.0 : 0)
            if rise > 0 && factor > 0 {
                len = max(len, (factor * rise) / Track.maxLaunchSlope)
            }
            let n = max(2, Int((len / Track.step).rounded()))
            for i in 1...n {
                let t = Double(i) / Double(n)
                let k: Double
                switch shape {
                case "kicker": k = pow(t, 1.3)
                case "linear": k = t
                default: k = t * t * (3 - 2 * t)
                }
                pts.append(base + rise * k)
            }
            base += rise
        }
        func pushWhoops(_ count: Int, _ amp: Double, _ spacing: Double) {
            // Peak slope of this profile is amp*pi/spacing; hold it to a face a
            // rider could actually ride rather than a wall.
            let amp = min(amp, (Track.maxFaceSlope * spacing) / .pi)
            let n = Int((Double(count) * spacing / Track.step).rounded())
            for i in 1...max(1, n) {
                let x = Double(i) * Track.step
                pts.append(base + amp * (1 - cos((x / spacing) * .pi * 2)) * 0.5)
            }
        }
        func pushRollers(_ count: Int, _ amp: Double, _ spacing: Double) {
            let amp = min(amp, (Track.maxFaceSlope * spacing) / .pi)
            let n = Int((Double(count) * spacing / Track.step).rounded())
            for i in 1...max(1, n) {
                let x = Double(i) * Track.step
                pts.append(base + amp * sin((x / spacing) * .pi * 2) * 0.5)
            }
        }
        func pushBerm(_ len: Double, _ depth: Double) {
            let n = max(2, Int((len / Track.step).rounded()))
            for i in 1...n {
                let t = Double(i) / Double(n)
                pts.append(base - depth * sin(t * .pi))
            }
        }
        func cursor() -> Double { Double(pts.count - 1) * Track.step }

        pushFlat(22)

        let jumpBias = biome.jumpBias * lerp(0.7, 1.35, d)
        let whoopBias = biome.whoopBias * lerp(0.6, 1.3, d)
        var lastKind = ""

        while cursor() < target {
            let table: [(String, Double)] = [
                ("tabletop", 1.0 * jumpBias),
                ("double", 0.85 * jumpBias),
                ("triple", 0.45 * jumpBias * d * 1.6),
                ("stepup", 0.6 * jumpBias),
                ("stepdown", 0.55 * jumpBias),
                ("whoops", 0.9 * whoopBias),
                ("rollers", 0.8),
                ("berm", 0.9),
                ("rhythm", 0.7 * jumpBias),
                ("straight", 0.5),
                ("hill", 0.7 * biome.rolling)
            ]
            func pick() -> String {
                let total = table.reduce(0.0) { $0 + $1.1 }
                var r = rnd.next() * total
                for (k, w) in table { r -= w; if r <= 0 { return k } }
                return "straight"
            }
            var kind = pick()
            if kind == lastKind && rnd.next() < 0.7 { kind = pick() }
            lastKind = kind

            let startX = cursor()
            let sev = lerp(0.7, 1.35, d) * rnd.range(0.85, 1.15)
            var risk = 0.2

            switch kind {
            case "tabletop":
                let h = rnd.range(1.1, 2.1) * sev
                let face = rnd.range(6.5, 9.5)
                pushRamp(face, h, "kicker")
                pushFlat(rnd.range(5, 12) * sev)
                pushRamp(face * 2.2, -h, "smooth")
                risk = 0.35
            case "double":
                let h = rnd.range(1.2, 2.0) * sev
                let gap = rnd.range(6, 10) * sev
                pushRamp(rnd.range(6, 8), h, "kicker")
                pushRamp(2.2, -h * 0.95, "linear")
                pushFlat(gap)
                pushRamp(rnd.range(4, 6), h * 0.95, "smooth")
                pushRamp(rnd.range(11, 15), -h, "smooth")
                risk = 0.65
            case "triple":
                let h = rnd.range(1.5, 2.4) * sev
                pushRamp(7.5, h, "kicker")
                pushRamp(2.4, -h, "linear")
                pushFlat(rnd.range(8, 11))
                pushRamp(4, h * 0.7, "smooth")
                pushRamp(6.5, -h * 0.7, "smooth")
                pushFlat(rnd.range(8, 11))
                pushRamp(4.5, h * 0.9, "smooth")
                pushRamp(14, -h * 0.9, "smooth")
                risk = 0.9
            case "stepup":
                let h = rnd.range(1.4, 2.6) * sev
                pushRamp(rnd.range(6.5, 8.5), h * 0.6, "kicker")
                pushRamp(2, -h * 0.15, "linear")
                pushFlat(rnd.range(5, 9))
                pushRamp(4, h * 0.55, "smooth")
                pushFlat(6)
                risk = 0.6
            case "stepdown":
                let h = rnd.range(1.3, 2.4) * sev
                pushRamp(7, h * 0.5, "kicker")
                pushRamp(1.6, -h * 0.1, "linear")
                pushFlat(rnd.range(6, 10))
                pushRamp(11, -h, "smooth")
                pushFlat(5)
                risk = 0.6
            case "whoops":
                let count = Int((rnd.range(5, 11) * lerp(0.8, 1.3, d)).rounded())
                pushWhoops(count, rnd.range(0.5, 1.0) * sev, rnd.range(3.4, 4.6))
                risk = 0.7
            case "rollers":
                pushRollers(Int(rnd.range(3, 6).rounded()), rnd.range(0.6, 1.4) * sev, rnd.range(6, 9))
                risk = 0.25
            case "berm":
                pushBerm(rnd.range(10, 18), rnd.range(0.8, 1.8) * sev)
                risk = 0.2
            case "rhythm":
                let n = Int(rnd.range(3, 5).rounded())
                let h = rnd.range(0.8, 1.5) * sev
                for _ in 0..<max(1, n) {
                    pushRamp(rnd.range(4.5, 6), h, "kicker")
                    pushRamp(rnd.range(7, 10), -h, "smooth")
                    pushFlat(rnd.range(1.5, 4))
                }
                risk = 0.8
            case "hill":
                let h = rnd.range(3, 9) * biome.rolling * 0.6
                pushRamp(rnd.range(14, 26), rnd.next() < 0.5 ? h : -h, "smooth")
                risk = 0.15
            default:
                pushFlat(rnd.range(10, 22))
                risk = 0.05
            }

            features.append(TrackFeature(kind: kind, x: startX, len: cursor() - startX, risk: risk))
            // Ramp the drift over the connector instead of stepping the base
            // and then running flat, which left a vertical curb at every
            // feature boundary.
            pushRamp(rnd.range(2, 6), (rnd.next() - 0.5) * 0.6 * biome.rolling, "smooth")
        }

        pushFlat(40)

        // Two light smoothing passes: raw sample joints read as invisible
        // curbs to the suspension solver.
        for _ in 0..<2 {
            if pts.count > 2 {
                for i in 1..<(pts.count - 1) {
                    pts[i] = pts[i] * 0.6 + (pts[i - 1] + pts[i + 1]) * 0.2
                }
            }
        }

        self.pts = pts
        self.features = features
        self.length = Double(pts.count - 1) * Track.step
        self.startX = 6
        self.finishX = self.length - 26

        var props: [Prop] = []
        var px: Double = 0
        while px < self.length {
            if rnd.next() < 0.35 {
                let kind: String
                if biome.key == "forest" { kind = "tree" }
                else if biome.stadium { kind = "banner" }
                else { kind = rnd.next() < 0.4 ? "flag" : "bale" }
                props.append(Prop(x: px + rnd.range(0, 4), kind: kind,
                                  scale: rnd.range(0.7, 1.4), back: rnd.next() < 0.5))
            }
            px += 6
        }
        self.props = props

        let prefix = ["Copper", "Iron", "Hollow", "Wild", "Broken", "Silver", "Red", "Black",
                      "High", "Thunder", "Dust", "Sky", "Storm", "Cinder", "Pine", "Salt"]
        let suffix = ["Ridge", "Basin", "Creek", "Flats", "Canyon", "Hollow", "Bowl", "Pass",
                      "Sands", "Bluff", "Gulch", "Park", "Valley", "Point", "Rise"]
        self.name = prefix[rnd.int(prefix.count)] + " " + suffix[rnd.int(suffix.count)]
    }

    func height(at x: Double) -> Double {
        let f = x / Track.step
        if f <= 0 { return pts[0] }
        if f >= Double(pts.count - 1) { return pts[pts.count - 1] }
        let i = Int(f)
        return lerp(pts[i], pts[i + 1], f - Double(i))
    }

    func slope(at x: Double) -> Double {
        return (height(at: x + Track.step) - height(at: x - Track.step)) / (2 * Track.step)
    }

    func feature(at x: Double) -> TrackFeature? {
        for f in features where x >= f.x && x < f.x + f.len { return f }
        return nil
    }

    /// Heavy landings dig a rut. Capped so a long session cannot erase a jump.
    func deform(at x: Double, depth: Double) {
        let i = Int((x / Track.step).rounded())
        for k in -3...3 {
            let j = i + k
            if j > 1 && j < pts.count - 1 {
                pts[j] -= depth * (1 - Double(abs(k)) / 4)
            }
        }
    }
}
